src/journal/journald-stream.c

```diff
@@ -548,9 +548,20 @@ static int stdout_stream_process(sd_event_source *es, int fd, uint32_t revents,
         if (ucred && ucred->pid != s->ucred.pid) {
                 /* force out any previously half-written lines from a different process, before we switch to
                  * the new ucred structure for everything we just added */
-                r = stdout_stream_scan(s, true);
-                if (r < 0)
-                        goto terminate;
+                if (s->length > 0) {
+                        char saved_char;
+
+                        saved_char = s->buffer[s->length];
+                        s->buffer[s->length] = '\0';
+
+                        r = stdout_stream_line(s, s->buffer, LINE_BREAK_EOF);
+                        if (r < 0)
+                                goto terminate;
+
+                        s->buffer[s->length] = saved_char;
+                        memmove(s->buffer, &s->buffer[s->length], l);
+                        s->length = 0;
+                }
```

Comment:
this looks wrong, the boolean parameter passed to stdout_stream_scan() called `force_flush` is precisely there to force out any pending data in the buffer regardless if terminated by some end of line character or not.

Reply:
Well the function stdout_stream_scan() was not "buggy", but did not handle the case where there was valid data after the data that needs to be flushed. This is why in the third commit I did refactor a little bit the code, to make it easier to understand

See https://github.com/systemd/systemd/pull/15685#discussion_r420028460